staging/src/k8s.io/apiserver/pkg/server/handler.go

```diff
@@ -98,6 +120,51 @@ func (a *APIServerHandler) ListedPaths() []string {
 	return handledPaths
 }
 
+type director struct {
+	name               string
+	goRestfulContainer *restful.Container
+	postGoRestfulMux   *mux.PathRecorderMux
+}
+
+func (d director) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	path := req.URL.Path
+
+	// we only need further inspection on paths under /apis
+	if !strings.HasPrefix(path, "/apis") {
+		d.goRestfulContainer.ServeHTTP(w, req)
+		return
+	}
+	// if we are exactly /apis or /apis/, pass it through since gorestful either handles discovery
+	// or if discovery isn't registered, it hasn't stolen /apis and it will fall through.
+	if path == "/apis" || path == "/apis/" {
+		d.goRestfulContainer.ServeHTTP(w, req)
+		return
+	}
+
+	// at this point we may or may not need to risk gorestful. We need to do it if there is a webservice root path
+	// other than /apis which is a prefix for our request.  Otherwise we skip it.
+	for _, ws := range d.goRestfulContainer.RegisteredWebServices() {
+		if ws.RootPath() == "/api" || ws.RootPath() == "/apis" || ws.RootPath() == "/" {
+			continue
+		}
+		if strings.HasPrefix(path, ws.RootPath()) {
```

Comment:
that's a lot of string allocations. (assuming all webservice rootpaths have no trailing slash) what about:

```go
if strings.HasPrefix(path, ws.RootPath()) {
  // ensure an exact match or a path boundary match
  if len(path) == len(ws.RootPath()) || path[len(ws.RootPath())] == '/' {
     ...
  }
}
```